package com.jakeporter.DocumentSummarizer.service;

import com.jakeporter.DocumentSummarizer.domainEntities.SummaryComponents;
import com.jakeporter.DocumentSummarizer.exceptions.*;
import com.jakeporter.DocumentSummarizer.utilities.fileUtils.FileInfoGetter;
import com.jakeporter.DocumentSummarizer.utilities.summarizers.PythonSummarizer;
import com.jakeporter.DocumentSummarizer.utilities.textExtractors.*;
import com.jakeporter.DocumentSummarizer.utilities.validators.PythonValidator;
import org.apache.commons.io.FilenameUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.io.InputStream;
import java.math.BigDecimal;
import java.math.RoundingMode;

@Service
public class FileService {

    Logger logger = LoggerFactory.getLogger(this.getClass());

    // had to turn off Spring's spring.servlet.multipart.max-file-size/max-request-size boundaries in application.properties and handle file size limits here (nothing else worked properly)
    public void validateFileSize(MultipartFile file) {
        final String FIVE_MB_IN_BYTES = "5242880";
        final String MB_IN_BYTES = "1048576";
        BigDecimal maxSize = new BigDecimal(FIVE_MB_IN_BYTES);
        BigDecimal mbInBytes = new BigDecimal(MB_IN_BYTES);
        BigDecimal fileSize = new BigDecimal(String.valueOf(file.getSize()));
        if (fileSize.compareTo(maxSize) > 0) {
            throw new FileTooLargeException("File size cannot exceed " + maxSize.divide(mbInBytes, RoundingMode.FLOOR) + " MB. Your file is " + fileSize.divide(mbInBytes, RoundingMode.FLOOR).setScale(2, RoundingMode.FLOOR) + " MB.");
        }
    }

    /* There's an ugly chain of exception throws in here- they're specific because I want to send different status codes to the client for each specific problem
       and I have to re-throw them in here for the ExceptionHandlerController to intercept them */
    public SummaryComponents summarize(MultipartFile mpFile, String language) {
        SummaryComponents components = null;
        FileType fileType;
        InputStream inputStream;
        try {
            // IDE says .getExtension() could return null, but it won't, as it throws an exception if it is
            fileType = new FileInfoGetter().getFileType(FilenameUtils.getExtension(mpFile.getOriginalFilename()));
        } catch (FileException e) {
            throw new FileException(e.getMessage());
        }
        try {
            inputStream = mpFile.getInputStream();
        } catch (IOException e) {
            e.printStackTrace();
            throw new FileException("Something went wrong processing the file.");
        }
        try {
            FileTextExtractor extractor = FileTextExtractorFactory.getExtractor(fileType);
            long start = System.currentTimeMillis();
            logger.info("FileService LINE 59: Text will now be extracted and summarized.");
            components = new PythonSummarizer(extractor, new PythonValidator()).summarize(inputStream, language);
            double executionTime = (System.currentTimeMillis() - start) / 1000D;
            logger.info("-- Execution time (sec): " + executionTime);
            logger.info("Summary components: " + components.toString());
        } catch (TextExtractorException e) {
            e.printStackTrace();
            throw new TextExtractorException(e.getMessage());
        } catch (SummaryException  | InterruptedException e) {
            e.printStackTrace();
            throw new SummaryException(e.getMessage());
        } catch (PythonScriptException e) {
            e.printStackTrace();
            throw new PythonScriptException(e.getMessage());
        } catch (Exception e) { // Don't know what uncaught exceptions could throw this at the moment, just want to ensure stack trace isn't shown to user
            e.printStackTrace();
            throw new SummaryException("Something unexpected went wrong, please tell Jake.");
        }
        return components;
    }

    public SummaryComponents summarize(String text, String language) {
        SummaryComponents components;
        try {
            components = new PythonSummarizer(new PythonValidator()).summarize(text, language);
        } catch (SummaryException | InterruptedException e) {
            e.printStackTrace();
            throw new SummaryException(e.getMessage());
        } catch (PythonScriptException e) {
            e.printStackTrace();
            throw new PythonScriptException(e.getMessage());
        } catch (Exception e) { // Don't know what uncaught exceptions could throw this at the moment, just want to ensure stack trace isn't shown to user
            e.printStackTrace();
            throw new SummaryException("Something unexpected went wrong, please tell Jake.");
        }
        return components;
    }
}
